### PR TITLE
✨ Autoscaling: scale down nodes

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?
This PR brings a mechanism to scale down the nodes created by the autoscaling application. (concerns only AWS deployments).

The following concepts were already known:
- A monitored node is a node with labels defined as in ```NODES_MONITORING_NODE_LABELS```
- An observed service is a service with labels defined as in ```NODES_MONITORING_SERVICE_LABELS``` (also its tasks are labeled the same)
- The autoscaling service scales up the cluster with fitting machines when it detects observed services that are pending resources (one at a time for now)
- The new EC2 instances are connected to the swarm in ```drain``` mode, labelled accordingly  and then made ```active``` so that the swarm can schedule the services on them

The following concepts are new:
- When a node is not running any observed service anymore, it is first set to ```drain``` (this will prevent the swarm from scheduling anything there)
- A drained node is kept around until the time since its start is above ```EC2_INSTANCES_TIME_BEFORE_TERMINATION``` (defaults to 55 minutes) modulo one hour. E.G. every started hour costs the same amount of money, therefore we can keep the instance around.
- If a drained node exceeds the ```EC2_INSTANCES_TIME_BEFORE_TERMINATION``` then it will be terminated and removed from the swarm
- If a new pending service appears, and there are still some drained nodes around that could handle the service, then they will be used before scaling up --> no need to wait 3 minutes in that case

### NOTE: this is not yet sending a message in rabbit MQ prior to termination. this will come later.

test coverage kept at around 99%

## Related issue/s

#3627

### CODE REVIEW

File: services/autoscaling/tests/unit/conftest.py [hunk 1]
```diff
@@ -296,70 +330,63 @@ async def _creator(
...
-                print(
-                    f"<-- service {found_service['Spec']['Name']} is now {service_task['Status']['State']} after {SUCCESS_STABLE_TIME_S} seconds"
-                )
+async def assert_for_service_state(
+    async_docker_client: aiodocker.Docker, service: Service, expected_states: list[str]
+) -> None:
+    SUCCESS_STABLE_TIME_S: Final[float] = 3
+    WAIT_TIME: Final[float] = 0.5
+    number_of_success = 0
+    async for attempt in AsyncRetrying(
```
Comment (pcrespov) [inline services/autoscaling/tests/unit/conftest.py:356]: MINOR: in this case, it feels better as decorator or ``assert_for_service_state``, right?

**Review by pcrespov — APPROVED** [review]
Great work!

File: services/autoscaling/tests/unit/conftest.py [hunk 2]
```diff
@@ -156,6 +165,17 @@ def app_settings(initialized_app: FastAPI) -> ApplicationSettings:
...
 
 
+@pytest.fixture
+def service_monitored_labels(
+    app_settings: ApplicationSettings,
+) -> dict[DockerLabelKey, str]:
+    assert app_settings.AUTOSCALING_NODES_MONITORING
+    return {
+        key: "true"
+        for key in app_settings.AUTOSCALING_NODES_MONITORING.NODES_MONITORING_SERVICE_LABELS
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/autoscaling/tests/unit/conftest.py:175]: well for now yes. but it is a label key

File: services/autoscaling/src/simcore_service_autoscaling/modules/ec2.py [hunk 3]
```diff
@@ -159,25 +170,69 @@ async def start_aws_instance(
...
-            # NOTE: this is currently deactivated as this makes starting an instance
-            # take between 2-4 minutes more and it seems to be responsive much before
-            # nevertheless if we get weird errors, this should be activated again!
-
-            # waiter = client.get_waiter("instance_status_ok")
-            # await waiter.wait(InstanceIds=[instance_id])
-            # logger.info("instance %s status is OK...", instance_id)
-
             # get the private IP
             instances = await self.client.describe_instances(InstanceIds=[instance_id])
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/autoscaling/src/simcore_service_autoscaling/modules/ec2.py:174]: there is a background task that will restart in any case. but that is true that this is not 100% wrapped in internal errors. not sure it is worth the effort for now.

File: services/autoscaling/src/simcore_service_autoscaling/modules/ec2.py [hunk 4]
```diff
@@ -159,25 +170,69 @@ async def start_aws_instance(
...
-            private_dns_name: str = instances["Reservations"][0]["Instances"][0][
-                "PrivateDnsName"
-            ]
+            instance = instances["Reservations"][0]["Instances"][0]
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/autoscaling/src/simcore_service_autoscaling/modules/ec2.py:175]: as everything runs inside the periodic task, it will be in worst case catched by the background task handler. then the task will restart.

File: services/autoscaling/tests/unit/conftest.py (hunk 1 above)
Comment (sanderegg) [reply services/autoscaling/tests/unit/conftest.py:356]: will check.. actually this is already a refactoring. fun.. originally I was using the retry all the time... and now you tell me to use it ;)

File: services/autoscaling/src/simcore_service_autoscaling/modules/ec2.py [hunk 5]
```diff
@@ -37,6 +40,14 @@ def _is_ec2_instance_running(instance: ReservationTypeDef):
     )
 
 
+@dataclass
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/autoscaling/src/simcore_service_autoscaling/modules/ec2.py:43]: are you sure this is python 3.9 compatible? I think only ```frozen``` is available

File: services/autoscaling/tests/unit/conftest.py (hunk 1 above)
Comment (sanderegg) [reply services/autoscaling/tests/unit/conftest.py:356]: so actually I want to wait for stabilisation. which is why I need to use it this way.

File: services/autoscaling/tests/unit/conftest.py [hunk 6]
```diff
@@ -103,6 +105,13 @@ def app_environment(
             "EC2_INSTANCES_SUBNET_ID": faker.pystr(),
             "EC2_INSTANCES_AMI_ID": faker.pystr(),
             "EC2_INSTANCES_ALLOWED_TYPES": json.dumps(ec2_instances),
+            "NODES_MONITORING_NODE_LABELS": json.dumps(["pytest.fake-node-label"]),
+            "NODES_MONITORING_SERVICE_LABELS": json.dumps(
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/autoscaling/tests/unit/conftest.py:109]: I did not fix it no.

File: services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py [hunk 7]
```diff
@@ -1,68 +1,203 @@
...
+    ):
+        if isinstance(ec2_instance_data, Ec2InstanceNotFoundError):
+            # skip if already terminating
+            continue
+        # NOTE: AWS price is hourly based (e.g. same price for a machine used 2 minutes or 1 hour, so we wait until 55 minutes)
+        elapsed_time_since_launched = (
+            datetime.utcnow().replace(tzinfo=timezone.utc)
+            - ec2_instance_data.launch_time
+        )
+        elapsed_time_since_full_hour = elapsed_time_since_launched % timedelta(hours=1)
```
Comment (mrnicegyu11) [inline services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py:121]: Will this work if the time changes to summer time?

Ridiculous question, I know, but let's try to avoid super-weird errors two times per year :D

I guess it should be fine since you explicitly use UTC

File: services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py [hunk 8]
```diff
@@ -1,68 +1,203 @@
...
+                get_ec2_client(app).terminate_instance(ec2_instance_data)
+                for _, ec2_instance_data in terminateable_nodes
+            )
+        )
+        logger.info(
+            "terminated the following machines: '%s'",
+            f"{[node.Description.Hostname for node,_ in terminateable_nodes if node.Description]}",
+        )
+        # since these nodes are being terminated, remove them from the swarm
+        await utils_docker.remove_nodes(
```
Comment (mrnicegyu11) [inline services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py:151]: What if this call fails? Is this somehow handled in a way that the autoscaling continuous in a proper way? Do we need an alert and human intervention in this case?
I have seen docker swarm struggle managing it's nodes, for example when we started an autoscaled machine fro a AMI image where the swarm was still joined, i.e. duplicating swarm members.

File: services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py [hunk 9]
```diff
@@ -107,13 +241,15 @@ async def check_dynamic_resources(app: FastAPI) -> None:
...
+                _EC2_INTERNAL_DNS_RE, new_instance_data.aws_private_dns
+            ):
                 new_instance_dns_name = match.group(1)
                 new_node = await utils_docker.wait_for_node(
-                    docker_client, new_instance_dns_name
+                    get_docker_client(app), new_instance_dns_name
                 )
                 await utils_docker.tag_node(
-                    docker_client,
+                    get_docker_client(app),
```
Comment (mrnicegyu11) [inline services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py:252]: if this call throws, is this handled in a nice way? I guess it could fail due to sheer statistical unluck, the docker swarm is a fragile system ;)

File: services/autoscaling/src/simcore_service_autoscaling/modules/ec2.py [hunk 10]
```diff
@@ -37,7 +40,15 @@ def _is_ec2_instance_running(instance: ReservationTypeDef):
     )
 
 
-@dataclass
+@dataclass(frozen=True)
+class EC2InstanceData:
+    launch_time: datetime.datetime
```
Comment (mrnicegyu11) [inline services/autoscaling/src/simcore_service_autoscaling/modules/ec2.py:45]: Maybe add a comment which TZ (UTC?) this here has for future reference

**Review by mrnicegyu11 — APPROVED** [review]
Very nice, thanks a lot

Some thoughts maybe for considerations :)

**Review by GitHK — APPROVED** [review]
Nice, looking forward for the part requiring coordination with the agent

File: services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py [hunk 11]
```diff
@@ -1,68 +1,203 @@
...
-    # 1. get monitored nodes information and resources
     docker_client = get_docker_client(app)
-    monitored_nodes = await utils_docker.get_monitored_nodes(
-        docker_client,
-        node_labels=app_settings.AUTOSCALING_NODES_MONITORING.NODES_MONITORING_NODE_LABELS,
+    active_empty_nodes = [
+        node
+        for node in monitored_nodes
+        if (
+            await utils_docker.compute_node_used_resources(
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py:36]: so the current use-case is as follow:
- osparc services (e.g. sim4life, ...) are marked as being special using the service_specifications table in the DB, therefore only these will get a special flag that bring them onto the scaled nodes
- from the docker documentation:
image
1. so this does not affect any container started as container (using docker run or docker-compose on the machine, which should anyway not be the case)
2. concerning the global services (such as cAdvisor), I am not quite sure yet whether the swarm actively stops them (I will try out and report back tomorrow). In the documentation they only say that replicated services are moved elsewhere.
3. if it kills the monitoring services, I think we might be able to live with it. at the moment I do not really see what alternative there is.
4. As to your question with ```utils_docker.compute_node_used_resources``` it checks for the "observed" services only, that is the ones that are specially marked (see the top bullet point here)

File: services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py [hunk 12]
```diff
@@ -1,68 +1,203 @@
...
+                docker_client,
+                node,
+                service_labels=app_settings.AUTOSCALING_NODES_MONITORING.NODES_MONITORING_SERVICE_LABELS,
+            )
+            == Resources.create_as_empty()
+        )
+        and (node.Spec is not None)
+        and (node.Spec.Availability == Availability.active)
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py:44]: so this is an interesting thought. What might happen that the docker swarm reports a node as down while it is still running?
1. networking problem within AWS after a successful initial connection
2. the machine crashed/hanged/non responsive
3. something else?

At the moment, the code will just remove any down node from the docker swarm nodes list, to prevent it from growing. it will not terminate the instance (with the current code that is). Now this is maybe a bad idea, I could also wait some time before removing it (such as checking the last updated field on the node, and maybe let an hour go through or similar).

Doing such a thing could maybe solve 1., 2. if someone manages to restart the broken machine within the timeout period. But as of now, that is beyond the scope of that service I think. we can brainstorm what might come as additional features.

File: services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py (hunk 7 above)
Comment (sanderegg) [reply services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py:121]: that is precisely why I use datetime.datetime, such that the time zone is set (default is UTC). And, for the case here what I check is anyway only how many minutes of a full hour elapsed.
from @Surfict I know that using an EC2 instance for 2 minutes or 59 minutes costs the same.
any started hour costs the same as the full hour. therefore I'm only interested in the number of minutes in each hour. So I do think that summer time, whatever time does not really matter here.

I will nevertheless add a test for negative hour difference to be on the same side.

File: services/autoscaling/src/simcore_service_autoscaling/modules/ec2.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (pcrespov) [reply services/autoscaling/src/simcore_service_autoscaling/modules/ec2.py]: aha... might be. we can check https://docs.python.org/3.9/library/dataclasses.html#dataclasses.dataclass

True only frozen available! 
OK, we need to move to 3.10 first thing next year :-))

File: services/autoscaling/tests/unit/conftest.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (pcrespov) [reply services/autoscaling/tests/unit/conftest.py:109]: mhh ok, let's check that tomorrow

File: services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py (hunk 8 above)
Comment (sanderegg) [reply services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py:151]: ok, so the problem you mention here I would describe as very unlikely to happen here. we did join a duplicate of a machine that was already in the swarm with exactly the same node ID... but let's say we do it:
- we have a swarm with nodes
- we take one of the node, and create an AMI out of it
- the autoscaling is started with that AMI
- it will then create nodes that have the same ID as another node already in the swarm
- the swarm will sometimes report the original node, sometimes the created one (as we saw together)
- then the autoscaling will create more nodes still using the same AMI, and they will never be added to the list, probably the swarm will eventually not work anymore at some point

Now what can I do from the autoscaling service point of view? 
1. I do not think I can double check that your AMI is not already connected to a swarm can I? 
2. Maybe I could run ```docker swarm leave``` as a preventive measure as instance bootup script , but it might also corrupt the swarm 
3. The best safeguard I could see, is that you have a test when the AMI is created...

If the call fails, the background task will catch it and restart the task.

File: services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py (hunk 9 above)
Comment (sanderegg) [reply services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py:252]: same as above, catched by the periodic task handler. then restarted 5 seconds later.

File: services/autoscaling/src/simcore_service_autoscaling/modules/ec2.py (hunk 10 above)
Comment (sanderegg) [reply services/autoscaling/src/simcore_service_autoscaling/modules/ec2.py:45]: the launch_time is filled up by the AWS SDK. therefore they define the time zone

File: services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py [hunk 13]
```diff
@@ -1,68 +1,203 @@
...
-        service_labels=app_settings.AUTOSCALING_NODES_MONITORING.NODES_MONITORING_SERVICE_LABELS,
-    )
-    await rabbitmq.post_state_message(
-        app,
-        monitored_nodes,
-        cluster_total_resources,
-        cluster_used_resources,
-        pending_tasks,
+    # get the corresponding ec2 instance data
+    # NOTE: some might be in the process of terminating and will not be found
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py:92]: the implications are that it is not necessary to terminate an instance twice.

File: services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py [hunk 14]
```diff
@@ -1,68 +1,203 @@
...
+        )
+        # since these nodes are being terminated, remove them from the swarm
+        await utils_docker.remove_nodes(
+            get_docker_client(app),
+            [node for node, _ in terminateable_nodes],
+            force=True,
+        )
+
+    # 3. we could ask on rabbit whether someone would like to keep that machine for something (like the agent for example), if that is the case, we wait another hour and ask again?
+    # 4.
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py:158]: probably, yes

File: services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py [hunk 15]
```diff
@@ -138,5 +274,51 @@ async def check_dynamic_resources(app: FastAPI) -> None:
...
                 "Task %s needs more resources than any EC2 instance "
                 "can provide with the current configuration. Please check.",
-                {f"{task.Name=}:{task.ServiceID=}"},
+                {
+                    f"{task.Name if task.Name else 'unknown task name'}:{task.ServiceID if task.ServiceID else 'unknown service ID'}"
+                },
             )
+
+
+async def check_dynamic_resources(app: FastAPI) -> None:
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py:283]: yes, this is a periodic task. this did not change since last time you checked.

File: services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py:36]: @mrnicegyu11:
so after a quick test this morning:
- after the node is set to drain: all services are removed including the monitoring services. So if you have an alternative idea to this let me know

File: services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py:44]: created a new issue out of this https://github.com/ITISFoundation/osparc-simcore/issues/3684

File: services/autoscaling/src/simcore_service_autoscaling/core/settings.py [hunk 16]
```diff
@@ -65,6 +65,11 @@ class EC2InstancesSettings(BaseCustomSettings):
         "this is required to start a new EC2 instance",
     )
 
+    EC2_INSTANCES_TIME_BEFORE_TERMINATION: datetime.timedelta = Field(
+        default=datetime.timedelta(minutes=55),
+        description="Defines the time an instance must be empty before the instance termination procedure may begin",
+    )
+
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/autoscaling/src/simcore_service_autoscaling/core/settings.py:72]: reworded

File: services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py [hunk 17]
```diff
@@ -1,68 +1,203 @@
...
+    if terminateable_nodes:
+        logger.info(
+            "the following nodes were found to be terminateable: '%s'",
+            f"{[node.Description.Hostname for node,_ in terminateable_nodes if node.Description]}",
+        )
+    return terminateable_nodes
+
+
+async def _try_scale_down_cluster(app: FastAPI, monitored_nodes: list[Node]) -> None:
+    # 2. once it is in draining mode and we are nearing a modulo of an hour we can start the termination procedure (parametrize this)
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py:137]: true

File: services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py (hunk 7 above)
Comment (sanderegg) [reply services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py:121]:
```python
datetime.timedelta(hours=-1, minutes=5) % datetime.timedelta(hours=1)
datetime.timedelta(seconds=300)
```

Comment (sanderegg) [reply services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py:121]: so it works
